Data/Recursive/Propagator/Seminaive.hs

```diff
@@ -0,0 +1,114 @@
+{-# LANGUAGE MultiParamTypeClasses, TypeFamilies, FlexibleContexts #-}
+module Data.Recursive.Propagator.Seminaive
+where
+
+import qualified Data.Set as S
+import Data.Set (Set)
+import Control.Monad (unless)
+import Control.Concurrent.MVar
+import Data.IORef
+-- import Data.Maybe
+
+import Data.POrder
+
+class (Eq a, Bottom a) => ChangeAction da a where
```

Comment:
Note that I replaced `Eq` with the check in `POrder`. This way we can deal with sets whose elements can’t be compared (it suffices to check the size of the sets).